core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala

```diff
@@ -49,29 +51,14 @@ private[spark] class JobWaiter[T](
     dagScheduler.cancelJob(jobId)
   }
 
-  override def taskSucceeded(index: Int, result: Any): Unit = synchronized {
-    if (_jobFinished) {
-      throw new UnsupportedOperationException("taskSucceeded() called on a finished JobWaiter")
-    }
+  override def taskSucceeded(index: Int, result: Any): Unit = {
```

Comment:
note that this does change behavior because `resultHandler` is no longer synchronized. Isn't this a potential problem, especially since the user can specify this?